src/network/mtp/impl.cpp

```diff
@@ -0,0 +1,1667 @@
+		if (packet->totaltime >= timeout)
+			count++;
+	}
+	return count;
+}
+
+std::vector<ConstSharedPtr<BufferedPacket>>
+	ReliablePacketBuffer::getResend(float timeout, u32 max_packets)
+{
+	MutexAutoLock listlock(m_list_mutex);
+	std::vector<ConstSharedPtr<BufferedPacket>> timed_outs;
+	for (auto &packet : m_list) {
+		// resend time scales exponentially with each cycle
+		const float pkt_timeout = timeout * powf(RESEND_SCALE_BASE, packet->resend_count);
+
+		if (packet->time < pkt_timeout)
+			continue;
+
+		// caller will resend packet so reset time and increase counter
+		packet->time = 0.0f;
+		packet->resend_count++;
+
+		timed_outs.emplace_back(packet);
+
+		if (timed_outs.size() >= max_packets)
+			break;
+	}
+	return timed_outs;
+}
+
+/*
+	IncomingSplitPacket
+*/
+
+bool IncomingSplitPacket::insert(u32 chunk_num, SharedBuffer<u8> &chunkdata)
+{
+	sanity_check(chunk_num < chunk_count);
+
+	// If chunk already exists, ignore it.
+	// Sometimes two identical packets may arrive when there is network
+	// lag and the server re-sends stuff.
+	if (chunks.find(chunk_num) != chunks.end())
+		return false;
+
+	// Set chunk data in buffer
+	chunks[chunk_num] = chunkdata;
+
+	return true;
+}
+
+SharedBuffer<u8> IncomingSplitPacket::reassemble()
+{
+	sanity_check(allReceived());
+
+	// Calculate total size
+	u32 totalsize = 0;
+	for (const auto &chunk : chunks)
+		totalsize += chunk.second.getSize();
+
+	SharedBuffer<u8> fulldata(totalsize);
+
+	// Copy chunks to data buffer
+	u32 start = 0;
+	for (u32 chunk_i = 0; chunk_i < chunk_count; chunk_i++) {
+		const SharedBuffer<u8> &buf = chunks[chunk_i];
+		memcpy(&fulldata[start], *buf, buf.getSize());
+		start += buf.getSize();
+	}
+
+	return fulldata;
+}
+
+/*
+	IncomingSplitBuffer
+*/
+
+IncomingSplitBuffer::~IncomingSplitBuffer()
+{
+	MutexAutoLock listlock(m_map_mutex);
+	for (auto &i : m_buf) {
+		delete i.second;
+	}
+}
+
+SharedBuffer<u8> IncomingSplitBuffer::insert(BufferedPacketPtr &p_ptr, bool reliable)
+{
+	MutexAutoLock listlock(m_map_mutex);
+	const BufferedPacket &p = *p_ptr;
+
+	u32 headersize = BASE_HEADER_SIZE + 7;
+	if (p.size() < headersize) {
+		errorstream << "Invalid data size for split packet" << std::endl;
+		return SharedBuffer<u8>();
+	}
+	u8 type = readU8(&p.data[BASE_HEADER_SIZE+0]);
+	u16 seqnum = readU16(&p.data[BASE_HEADER_SIZE+1]);
+	u16 chunk_count = readU16(&p.data[BASE_HEADER_SIZE+3]);
+	u16 chunk_num = readU16(&p.data[BASE_HEADER_SIZE+5]);
+
+	if (type != PACKET_TYPE_SPLIT) {
+		errorstream << "IncomingSplitBuffer::insert(): type is not split"
+			<< std::endl;
+		return SharedBuffer<u8>();
+	}
+	if (chunk_num >= chunk_count) {
+		errorstream << "IncomingSplitBuffer::insert(): chunk_num=" << chunk_num
+				<< " >= chunk_count=" << chunk_count << std::endl;
+		return SharedBuffer<u8>();
+	}
+
+	// Add if doesn't exist
+	IncomingSplitPacket *sp;
+	if (m_buf.find(seqnum) == m_buf.end()) {
+		sp = new IncomingSplitPacket(chunk_count, reliable);
```

Comment:
Old code, new question: Is this ever freed if we return early?

Reply:
doesn't look like it